ugvc/mrd/bqsr_inference_utils.py

```diff
@@ -0,0 +1,205 @@
+from __future__ import annotations
+
+import os
+import subprocess
+from os.path import join as pjoin
+
+import joblib
+import numpy as np
+import pandas as pd
+import pysam
+from joblib import Parallel, delayed
+from tqdm import tqdm
+
+from ugvc.dna.utils import revcomp
+from ugvc.utils.cloud_sync import cloud_sync
+from ugvc.utils.misc_utils import set_pyplot_defaults
+from ugvc.vcfbed.variant_annotation import get_cycle_skip_dataframe, get_motif_around
+
+set_pyplot_defaults()
+
+
+def iterate_vcf_in_chunks(vcf_file, chunk_size, model, output_file, chrstr, params, show_progress=True):
+    print(f"fetching chr {chrstr} from {vcf_file}\n")
+    with pysam.VariantFile(vcf_file) as vcf_in, pysam.VariantFile(output_file, "w", header=vcf_in.header) as vcf_out:
+
+        vcf_in.header.add_meta(
+            "INFO", items=[("ID", "ML_QUAL"), ("Number", 1), ("Type", "Float"), ("Description", "xgboost model score")]
+        )
+
+        # Add a comment to the output file indicating the use of an ML model
+        vcf_out.header.add_line(
+            '##INFO=<ID=ML_QUAL,Number=1,Type=Float,Description="Quality assigned using an ML model">'
+        )
+
+        # Calculate the total number of chunks
+        # total_variants = vcf_in.count_records()
+        # total_chunks = (total_variants - 1) // chunk_size + 1
+
+        # Iterate over the variants in chunks
+        variants = vcf_in.fetch(chrstr)
+        iterator = tqdm(total=None, disable=not show_progress)
+        chunk = [None] * chunk_size
+        index = 0
+        for variant in variants:
+            chunk[index] = variant
+            index += 1
+
+            if index == chunk_size:
+                process_chunk(chunk, model, vcf_out, params)
+                chunk = [None] * chunk_size
+                index = 0
+                iterator.update(1)
+
+        # Process the remaining variants (if any) in the last chunk
+        if index > 0:
+            process_chunk(chunk[:index], model, vcf_out, params)
+            iterator.update(1)
+
+        iterator.close()
+
+
+def process_chunk(chunk, model, vcf_out, params):
+
+    # Convert variants to a pandas DataFrame
```

Comment:
We definitely should, and seeing this now I realize the balanced ePCR tags should be handled in one dedicated place. This is something we should discuss because there are hanging changes in another branch.